agents/rl_agent.c

```diff
@@ -95,20 +82,37 @@ int get_action_exploit(char *table, td_agent_t *agent)
     return max_act;
 }
 
-int play_td(char *table, td_agent_t *agent)
+int play_rl(char *table, rl_agent_t *agent)
 {
     int move = get_action_exploit(table, agent);
     table[move] = agent->player;
     return move;
 }
 
-void update_state_value(char *cur_state, char *nxt_state, td_agent_t *agent)
+void store_state_value(rl_agent_t *agent, unsigned int N_STATES)
 {
-    int cur_state_hash = table_to_hash(cur_state);
-    int nxt_state_hash = table_to_hash(nxt_state);
-    int nxt_reward = get_score(hash_to_table(nxt_state_hash), agent->player);
-    agent->state_value[cur_state_hash] =
-        (1 - LEARNING_RATE) * agent->state_value[cur_state_hash] +
-        LEARNING_RATE *
-            (nxt_reward + GAMMA * agent->state_value[nxt_state_hash]);
+    FILE *fptr = NULL;
+    if ((fptr = fopen(MODEL_NAME, "wb")) == NULL) {
+        perror("Failed to open file");
+        exit(1);
+    }
+    if (fwrite(agent[0].state_value, N_STATES * sizeof(float), 1, fptr) != 1) {
+        perror("Failed to write file");
+        exit(1);
+    }
+    if (fwrite(agent[1].state_value, N_STATES * sizeof(float), 1, fptr) != 1) {
+        perror("Failed to write file");
+        exit(1);
+    }
+    fclose(fptr);
 }
+
+void init_rl_agent(rl_agent_t *agent, unsigned int state_num, char player)
+{
+    agent->player = player;
+    agent->state_value = malloc(sizeof(float) * state_num);
+    if (!(agent->state_value)) {
+        perror("Failed to allocate memory");
+        exit(1);
+    }
+}
```

Comment:
Always ensure that all of your text files end with a newline.

See: https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e